### PR TITLE
Adagrad Optimizer Implementation

### PR DESCRIPTION
**Pull request by Spnetic-5**
Reference: [PyTorch Docs]( https://pytorch.org/docs/stable/generated/torch.optim.Adagrad.html#torch.optim.Adagrad)

### CODE REVIEW

File: src/nf/nf_optimizers.f90 [hunk 1]
```diff
@@ -87,6 +87,20 @@ end subroutine minimize
...
 
+  type, extends(optimizer_base_type) :: adagrad
+    !! Adagrad optimizer by Duchi et al. (2011)
+    !!
+    !! Duchi, J., Hazan, E. and Singer, Y., 2011. Adaptive subgradient
+    !! methods for online learning and stochastic optimization. Journal
+    !! of Machine Learning Research, 12(Jul), pp.2121-2159.
+    !! http://www.jmlr.org/papers/volume12/duchi11a/duchi11a.pdf
+    real :: epsilon = 1e-8
+    real, allocatable, private :: sum_squared_gradient(:)
```
Comment (milancurcic) [inline src/nf/nf_optimizers.f90:98]: Need `learning_rate_decay` here ($\eta$ in https://pytorch.org/docs/stable/generated/torch.optim.Adagrad.html#torch.optim.Adagrad)

File: src/nf/nf_optimizers.f90 [hunk 2]
```diff
@@ -187,10 +201,34 @@ pure subroutine minimize_adam(self, param, gradient)
     ! Update parameters.
     param = param &
       - self % learning_rate * m_hat / (sqrt(v_hat) + self % epsilon) &
-      - self % weight_decay_decoupled * param
+      - self % learning_rate * self % weight_decay_decoupled * param
```
Comment (milancurcic) [inline src/nf/nf_optimizers.f90:204]: I suggest:

```
param = param &
  - self % learning_rate * (m_hat / (sqrt(v_hat) + self % epsilon) + self % weight_decay_decoupled * param)
```

File: src/nf/nf_optimizers.f90 [hunk 3]
```diff
@@ -187,10 +201,34 @@ pure subroutine minimize_adam(self, param, gradient)
...
+    !! Concrete implementation of an Adagrad optimizer update rule.
+    class(adagrad), intent(inout) :: self
+    real, intent(inout) :: param(:)
+    real, intent(in) :: gradient(:)
+
+    ! Update the sum of squared gradients using the Adagrad rule
+    self % sum_squared_gradient = self % sum_squared_gradient + gradient**2
+
+    ! Update the network parameters based on the new sum of squared gradients
+    param = param - self % learning_rate &
```
Comment (milancurcic) [inline src/nf/nf_optimizers.f90:229]: The learning rate needs to adapt with every time step here, according to https://pytorch.org/docs/stable/generated/torch.optim.Adagrad.html#torch.optim.Adagrad.

File: src/nf/nf_optimizers.f90 (hunk 1 above)
Comment (milancurcic) [reply src/nf/nf_optimizers.f90:98]: Both Keras and PyTorch implement the `weight_decay` (the L2 kind) in Adagrad. We may as well do that since we're already here.

**Comment by milancurcic** [comment]
Thanks @Spnetic-5. I believe it's correct now. In your original implementation, the L2 regularization was not accounted for in the accumulation of the squared gradients because you applied it later in the param update. The learning rate decay was also doubly accounted for because in each step the learning rate should be amortized relative to the original learning rate, not the one from the previous step. Subtle differences that weren't caught in the tests.

I'll go ahead and merge, please release v0.15.0 when you get a chance.
